Add OpenJDK as a dependency again

The removal of openjdk in #305 introduced some unintended side-effects of removing openjdk8 in excell files:

| before | after |
|----|----|
| ![xlsxbefore](https://user-images.githubusercontent.com/47065258/214285368-bff40f3e-8c28-4b6f-8d11-95f7e342b0d1.png) | ![xlsxafter](https://user-images.githubusercontent.com/47065258/214285392-c124bab7-832d-4ada-b663-2423fb292120.png)  |

We missed these because we didn't manually inspect all the output files and also because the conversion process dones't fail as soon as it gets an unexpected JSON from container:

```
Converting document to safe PDF
[DEBUG] Marking doc GQN_Qw as 'converting'
[INFO ] > /usr/bin/podman run --network none -u dangerzone --security-opt no-new-privileges --userns keep-id --cap-drop all --rm -v /home/user/dangerzone/tests/test_docs/sample.xlsx:/tmp/input_file -v /home/user/.config/dangerzone/tmp/tmpphanvmpz/pixels:/dangerzone dangerzone.rocks/dangerzone /usr/bin/python3 /usr/local/bin/dangerzone.py document-to-pixels
[INFO ] [doc GQN_Qw] 0% Converting to PDF using LibreOffice
[ERROR] Invalid JSON returned from container:

	 javaldx: Could not find a Java Runtime Environment!

[ERROR] Invalid JSON returned from container:

	 Warning: failed to read path from javaldx

[ERROR] Invalid JSON returned from container:

	 convert /tmp/input_file -> /tmp/input_file.pdf using filter : calc_pdf_Export

[INFO ] [doc GQN_Qw] 3% Calculating number of pages
[INFO ] [doc GQN_Qw] 10% Converting page 1/6 to pixels
[INFO ] [doc GQN_Qw] 18% Converting page 2/6 to pixels
[INFO ] [doc GQN_Qw] 25% Converting page 3/6 to pixels
[INFO ] [doc GQN_Qw] 33% Converting page 4/6 to pixels
[INFO ] [doc GQN_Qw] 40% Converting page 5/6 to pixels
[INFO ] [doc GQN_Qw] 48% Converting page 6/6 to pixels
```
It needs to be reverted. And out tests should at least include some sort of diff so things like these get flagged.

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -151,10 +151,10 @@ jobs:
       - run:
           name: Install poetry dependencies
           command: |
-            # Remove this pin once the upstream Poetry issue is fixed:
+            sudo pip3 install poetry
+            # This flag is important, due to an open upstream Poetry issue:
             # https://github.com/python-poetry/poetry/issues/7184
-            sudo pip3 install poetry==1.2.2
-            poetry install
+            poetry install --no-ansi
       - run:
           name: Prepare cache directory
           command: |

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -156,6 +156,10 @@ jobs:
             # https://github.com/python-poetry/poetry/issues/7184
             poetry install --no-ansi
       - run:
+          name: Install test dependencies
+          command: |
+            sudo apt-get install -y libqt5gui5 --no-install-recommends
+      - run:
           name: Prepare cache directory
           command: |
             sudo mkdir -p /caches

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -223,6 +223,32 @@ jobs:
             ./dev_scripts/env.py --distro ubuntu --version 22.04 run --dev \
                 bash -c 'cd dangerzone; poetry run make test'
 
+  ci-fedora-37:
+    machine:
+      image: ubuntu-2004:202111-01
+    steps:
+      - checkout
+      - run: *install-podman
+
+      - run:
+          name: Prepare cache directory
+          command: |
+            sudo mkdir -p /caches
+            sudo chown -R $USER:$USER /caches
+      - restore_cache: *restore-cache
+      - run: *copy-image
+
+      - run:
+          name: Prepare Dangerzone environment
+          command: |
+            ./dev_scripts/env.py --distro fedora --version 37 build-dev
+
+      - run:
+          name: Run CI tests
+          command: |
+            ./dev_scripts/env.py --distro fedora --version 37 run --dev \
+                bash -c 'cd dangerzone; poetry run make test'
+
   ci-fedora-36:
     machine:
       image: ubuntu-2004:202111-01
@@ -464,6 +490,9 @@ workflows:
           requires:
             - build-container-image
       - ci-debian-bookworm:
+          requires:
+            - build-container-image
+      - ci-fedora-37:
           requires:
             - build-container-image
       - ci-fedora-36:

--- a/container/Dockerfile
+++ b/container/Dockerfile
@@ -6,6 +6,7 @@ RUN apk -U upgrade && \
     ghostscript \
     graphicsmagick \
     libreoffice \
+    openjdk8 \
     poppler-utils \
     python3 \
     py3-magic \

--- a/container/dangerzone.py
+++ b/container/dangerzone.py
@@ -12,6 +12,7 @@ pixels_to_pdf:
 - 95%-100%: Compress the final PDF
 """
 
+import asyncio
 import glob
 import json
 import os
@@ -20,7 +21,7 @@ import shutil
 import subprocess
 import sys
 import time
-from typing import Callable, Dict, List, Optional, Union
+from typing import Callable, Dict, List, Optional, Tuple, Union
 
 import magic
 
@@ -31,7 +32,30 @@ DEFAULT_TIMEOUT: float = 120
 COMPRESSION_TIMEOUT: float = 10
 
 
-def run_command(
+async def read_stream(sr: asyncio.StreamReader, callback: Callable = None) -> bytes:
+    """Consume a byte stream line-by-line.
+
+    Read all lines in a stream until EOF. If a user has passed a callback, call it for
+    each line.
+
+    Note that the lines are in bytes, since we can't assume that all command output will
+    be UTF-8 encoded. Higher level commands are advised to decode the output to Unicode,
+    if they know its encoding.
+    """
+    buf = b""
+    while True:
+        line = await sr.readline()
+        if sr.at_eof():
+            break
+        if callback is not None:
+            callback(line)
+        # TODO: This would be a good place to log the received line, mostly for debug
+        # logging.
+        buf += line
+    return buf
+
+
+async def run_command(
     args: List[str],
     *,
     error_message: str,
@@ -39,68 +63,52 @@ def run_command(
     timeout: float = DEFAULT_TIMEOUT,
     stdout_callback: Callable = None,
     stderr_callback: Callable = None,
-) -> None:
-    """
-    Runs a command and returns the result.
+) -> Tuple[bytes, bytes]:
+    """Run a command and get its output.
+
+    Run a command using asyncio.subprocess, consume its standard streams, and return its
+    output in bytes.
 
     :raises RuntimeError: if the process returns a non-zero exit status
     :raises TimeoutError: if the process times out
     """
-    if stdout_callback is None and stderr_callback is None:
-        try:
-            subprocess.run(args, timeout=timeout, check=True)
-        except subprocess.CalledProcessError as e:
-            raise RuntimeError(error_message) from e
-        except subprocess.TimeoutExpired as e:
-            raise TimeoutError(timeout_message) from e
+    # Start the provided command, and return a handle. The command will run in the
+    # background.
+    proc = await asyncio.subprocess.create_subprocess_exec(
+        *args,
+        stdout=asyncio.subprocess.PIPE,
+        stderr=asyncio.subprocess.PIPE,
+    )
 
-    else:
-        p = subprocess.Popen(
-            args,
-            stdout=subprocess.PIPE,
-            stderr=subprocess.PIPE,
-            universal_newlines=True,
-        )
+    assert proc.stdout is not None
+    assert proc.stderr is not None
 
-        # Progress callback requires a manually implemented timeout
-        start_time = time.time()
+    # Create asynchronous tasks that will consume the standard streams of the command,
+    # and call callbacks if necessary.
+    stdout_task = asyncio.create_task(read_stream(proc.stdout, stdout_callback))
+    stderr_task = asyncio.create_task(read_stream(proc.stderr, stderr_callback))
 
-        # Make reading from stdout or stderr non-blocking
-        if p.stdout:
-            os.set_blocking(p.stdout.fileno(), False)
-        if p.stderr:
-            os.set_blocking(p.stderr.fileno(), False)
+    # Wait until the command has finished, for a specific timeout. Then, verify that the
+    # command has completed successfully. In any other case, raise an exception.
+    try:
+        ret = await asyncio.wait_for(proc.wait(), timeout=timeout)
+    except asyncio.exceptions.TimeoutError:
+        raise TimeoutError(timeout_message)
+    if ret != 0:
+        raise RuntimeError(error_message)
 
-        while True:
-            # Processes hasn't finished
-            if p.poll() is not None:
-                if p.returncode != 0:
-                    raise RuntimeError(error_message)
-                break
-
-            # Check if timeout hasn't expired
-            if time.time() - start_time > timeout:
-                p.kill()
-                raise TimeoutError(timeout_message)
-
-            if p.stdout and stdout_callback is not None:
-                line = p.stdout.readline()
-                if len(line) > 0:
-                    line = line.rstrip()  # strip trailing "\n"
-                    stdout_callback(line)
-
-            if p.stderr and stderr_callback is not None:
-                line = p.stderr.readline()
-                if len(line) > 0:
-                    line = line.rstrip()  # strip trailing "\n"
-                    stderr_callback(line)
+    # Wait until the tasks that consume the command's standard streams have exited as
+    # well, and return their output.
+    stdout = await stdout_task
+    stderr = await stderr_task
+    return (stdout, stderr)
 
 
 class DangerzoneConverter:
     def __init__(self) -> None:
         self.percentage: float = 0.0
 
-    def document_to_pixels(self) -> None:
+    async def document_to_pixels(self) -> None:
 
         conversions: Dict[str, Dict[str, Optional[str]]] = {
             # .pdf
@@ -194,7 +202,7 @@ class DangerzoneConverter:
                 "/tmp",
                 "/tmp/input_file",
             ]
-            run_command(
+            await run_command(
                 args,
                 error_message="Conversion to PDF with LibreOffice failed",
                 timeout_message=f"Error converting document to PDF, LibreOffice timed out after {DEFAULT_TIMEOUT} seconds",
@@ -208,7 +216,7 @@ class DangerzoneConverter:
                 "/tmp/input_file",
                 "/tmp/input_file.pdf",
             ]
-            run_command(
+            await run_command(
                 args,
                 error_message="Conversion to PDF with GraphicsMagick failed",
                 timeout_message=f"Error converting document to PDF, GraphicsMagick timed out after {DEFAULT_TIMEOUT} seconds",
@@ -222,25 +230,21 @@ class DangerzoneConverter:
 
         # Obtain number of pages
         self.update_progress("Calculating number of pages")
-        self.num_pages: Union[None, int] = None
-
-        def get_num_pages(line: str) -> None:
-            search = re.search(r"^Pages:           (\d+)", line)
-            if search is not None:
-                self.num_pages = int(search.group(1))
-
-        run_command(
+        stdout, _ = await run_command(
             ["pdfinfo", pdf_filename],
             error_message="PDF file is corrupted",
             timeout_message=f"Extracting metadata from PDF timed out after 1 second",
             timeout=1,
-            stdout_callback=get_num_pages,
         )
-        if self.num_pages == None:
-            raise ValueError("Number of pages could not be extraced from PDF")
 
-        def pdftoppm_progress_callback(line: str) -> None:
-            """Function called for every line the 'pdftoppm'command outputs
+        search = re.search(r"Pages:\s*(\d+)\s*\n", stdout.decode())
+        if search is not None:
+            self.num_pages: int = int(search.group(1))
+        else:
+            raise ValueError("Number of pages could not be extracted from PDF")
+
+        def pdftoppm_progress_callback(line: bytes) -> None:
+            """Function called for every line the 'pdftoppm' command outputs
 
             Sample pdftoppm output:
 
@@ -253,7 +257,7 @@ class DangerzoneConverter:
             Each successful line is in the format "{page} {page_num} {ppm_filename}"
             """
             try:
-                (page_str, num_pages_str, _) = line.split()
+                (page_str, num_pages_str, _) = line.decode().split()
                 num_pages = int(num_pages_str)
                 page = int(page_str)
             except ValueError as e:
@@ -305,8 +309,8 @@ class DangerzoneConverter:
         page_base = "/tmp/page"
 
         # Convert to PPM, which is essentially an RGB format
-        pdftoppm_timeout = 1.0 * self.num_pages  # type: ignore [operator]
-        run_command(
+        pdftoppm_timeout = 1.0 * self.num_pages
+        await run_command(
             [
                 "pdftoppm",
                 pdf_filename,
@@ -329,7 +333,7 @@ class DangerzoneConverter:
         ):
             shutil.move(filename, "/dangerzone")
 
-    def pixels_to_pdf(self) -> None:
+    async def pixels_to_pdf(self) -> None:
         self.percentage = 50.0
 
         num_pages = len(glob.glob("/dangerzone/page-*.rgb"))
@@ -354,7 +358,7 @@ class DangerzoneConverter:
                 self.update_progress(
                     f"Converting page {page}/{num_pages} from pixels to searchable PDF"
                 )
-                run_command(
+                await run_command(
                     [
                         "gm",
                         "convert",
@@ -368,7 +372,7 @@ class DangerzoneConverter:
                     error_message=f"Page {page}/{num_pages} conversion to PNG failed",
                     timeout_message=f"Error converting pixels to PNG, convert timed out after {DEFAULT_TIMEOUT} seconds",
                 )
-                run_command(
+                await run_command(
                     [
                         "tesseract",
                         png_filename,
@@ -387,7 +391,7 @@ class DangerzoneConverter:
                 self.update_progress(
                     f"Converting page {page}/{num_pages} from pixels to PDF"
                 )
-                run_command(
+                await run_command(
                     [
                         "gm",
                         "convert",
@@ -410,7 +414,7 @@ class DangerzoneConverter:
         for page in range(1, num_pages + 1):
             args.append(f"/tmp/page-{page}.pdf")
         args.append(f"/tmp/safe-output.pdf")
-        run_command(
+        await run_command(
             args,
             error_message="Merging pages into a single PDF failed",
             timeout_message=f"Error merging pages into a single PDF, pdfunite timed out after {DEFAULT_TIMEOUT} seconds",
@@ -421,7 +425,7 @@ class DangerzoneConverter:
         # Compress
         self.update_progress("Compressing PDF")
         compress_timeout = num_pages * COMPRESSION_TIMEOUT
-        run_command(
+        await run_command(
             ["ps2pdf", "/tmp/safe-output.pdf", "/tmp/safe-output-compressed.pdf"],
             timeout_message=f"Error compressing PDF, ps2pdf timed out after {compress_timeout} seconds",
             error_message="Compressing PDF failed",
@@ -444,7 +448,7 @@ class DangerzoneConverter:
         sys.stdout.flush()
 
 
-def main() -> int:
+async def main() -> int:
     if len(sys.argv) != 2:
         print(f"Usage: {sys.argv[0]} [document-to-pixels]|[pixels-to-pdf]")
         return -1
@@ -453,9 +457,9 @@ def main() -> int:
 
     try:
         if sys.argv[1] == "document-to-pixels":
-            converter.document_to_pixels()
+            await converter.document_to_pixels()
         elif sys.argv[1] == "pixels-to-pdf":
-            converter.pixels_to_pdf()
+            await converter.pixels_to_pdf()
     except (RuntimeError, TimeoutError, ValueError) as e:
         converter.update_progress(str(e), error=True)
         return 1
@@ -464,4 +468,4 @@ def main() -> int:
 
 
 if __name__ == "__main__":
-    sys.exit(main())
+    sys.exit(asyncio.run(main()))

--- a/dev_scripts/env.py
+++ b/dev_scripts/env.py
@@ -74,7 +74,7 @@ RUN dnf install -y rpm-build podman python3 make python3-pip qt5-qtbase-gui \
 
 # FIXME: Drop this fix after it's resolved upstream.
 # See https://github.com/freedomofpress/dangerzone/issues/286#issuecomment-1347149783
-RUN dnf reinstall -y shadow-utils && dnf clean all
+RUN rpm --restore shadow-utils
 
 RUN dnf install -y mupdf && dnf clean all
 """
@@ -120,7 +120,7 @@ RUN dnf install -y mupdf && dnf clean all
 
 # FIXME: Drop this fix after it's resolved upstream.
 # See https://github.com/freedomofpress/dangerzone/issues/286#issuecomment-1347149783
-RUN dnf reinstall -y shadow-utils && dnf clean all
+RUN rpm --restore shadow-utils
 """
 
 # The Dockerfile for building an environment with Dangerzone installed in it. Parts of

--- a/dev_scripts/pytest-wrapper.py
+++ b/dev_scripts/pytest-wrapper.py
@@ -14,7 +14,6 @@ import re
 import subprocess
 import sys
 
-import pytest
 from pkg_resources import parse_version
 
 from dangerzone.isolation_provider.container import Container
@@ -30,14 +29,22 @@ def get_podman_version():
     return version.split("-dev")[0]  # exclude "-dev" suffix from version
 
 
+def run_tests(pytest_args):
+    cmd = ["pytest"] + pytest_args
+    try:
+        subprocess.run(cmd, check=True)
+    except subprocess.CalledProcessError:
+        sys.exit(1)
+
+
 def run_tests_in_parallel(pytest_args):
-    args = pytest_args + ["-n", "4"]
-    exit_code = pytest.main(args)
+    print("running tests in parallel")
+    run_tests(pytest_args + ["-n", "4"])
 
 
 def run_tests_in_sequence(pytest_args):
     print("running tests sequentially")
-    exit_code = pytest.main(pytest_args)
+    run_tests(pytest_args)
 
 
 if __name__ == "__main__":

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -6,11 +6,10 @@ authors = ["Micah Lee <micah.lee@theintercept.com>"]
 license = "MIT"
 
 [tool.poetry.dependencies]
-python = ">=3.7,<3.11"
+python = ">=3.7,<3.12"
 click = "*"
 appdirs = "*"
-PySide2 = {version = "5.15.2.1", platform = "linux"}
-PySide6 = {version = "^6.4.1", markers = "sys_platform == 'win32' or sys_platform == 'darwin'"}
+PySide6 = "^6.4.1"
 colorama = "*"
 pyxdg = {version = "*", platform = "linux"}
 

